### ISSUE TITLE
FEM for 3D Tet mesh

### ISSUE THREAD

**Issue by EmJay276** [sgsellan/gpytoolbox#46]
Are you interested in me implementing the FEM for 3D Tets?

I already have a working version for every element type using [gmsh](https://gmsh.info/doc/texinfo/gmsh.html).
You probably want to stay with numpy / scipy only, correct?
For linear elastic tet elements I can adapt it to only use numpy/scipy. If you don't need it next week 😀.

- Can you give me some hints where to find existing code/functionality for 3D Tets in gpytoolbox?

- Do you use voigt notation for stresses/straits? 
![grafik](https://user-images.githubusercontent.com/8619525/211308485-4d9f5143-b515-46d3-b0c0-983c94f105bc.png)
- Or diagonal notation for stresses/straits?
![grafik](https://user-images.githubusercontent.com/8619525/211309119-f38f5279-3778-4c67-8771-ec485ffcb882.png)
(only relevant for anisotropic materials)

**Comment by sgsellan** [comment]
Oh, and I forgot to answer one of your questions: yes, we would like everything to be numpy/scipy-based (like our triangle mesh FEM implementations).

**Comment by sgsellan** [comment]
Indeed, there is currently no `read_mesh` and `write_mesh` for 3D elements (@odedstein may contradict me in case this is not true).

I don't think we have any powerful reason to restrict ourselves to linear, tetrahedral elements, but for the fact that they make life much easier. Thus if I were writing this, I would start by doing linear tets, then quadratic tets, then other elements. As long as (like you say) the user can easily know what they're calling and it has a reasonable default, I have no opposition to the code being richer and would be happy to merge it.

**Comment by EmJay276** [comment]
Ok thanks, then I will write the methods to make them easily extendable and annotate there it can be done.

**Comment by sgsellan** [comment]
Sounds good! Looking forward to it :-)

**Comment by EmJay276** [comment]
@sgsellan @odedstein 
I did some research on 3D mesh formats. I didn't find any commonly used file format (please correct me if I'm wrong).

I would suggest to use the gmsh `.msh` format ( https://gmsh.info/doc/texinfo/gmsh.html#Gmsh-file-formats ).
It seems well documented and if I keep my data structure compatible with the gmsh node/element structure, I can reuse most of my existing code.
It would also be much easier to extend to different element shapes and types.

Does this sound reasonable to you?

**Comment by odedstein** [comment]
I like the gmsh .msh format. I think it's a reasonble choice.

**Comment by EmJay276** [comment]
Just an update for you, im currently working on the gmsh importer (works fine already). I hadn't much spare time recently.
